### ISSUE TITLE
Stuck at Establishing a Connection

### ISSUE THREAD

**Issue by yusing** [DomiStyle/docker-tor-browser#53]
![image](https://github.com/DomiStyle/docker-tor-browser/assets/23052124/ca1dfd81-6dbb-40af-852d-9333b5d4fb2d)

**Comment by lawndoc** [comment]
Hi @yusing, thanks for reaching out. If you haven't already, please try to close all tabs and wait a moment -- this will restart the browser. 

I've had issues in the past with the browser staying open for too long making it unable to connect. As far as I can tell, it's an issue with the Tor Browser itself and nothing we can do on our end. Restarting the process as I described above fixes it for me.

If restarting the Tor Browser doesn't fix the issue for you, then the problem is probably related to your network. Please refer to the Tor project's [official network troubleshooting documentation](https://tb-manual.torproject.org/troubleshooting/) for help with this.

Another troubleshooting step you can take: try running the Tor Browser locally on the same network that your container is running on. Do you get the same issue when establishing a connection, or does it work outside of the container?

**Comment by lawndoc** [comment]
@yusing have you had a chance to try any of the above suggestions? If we don't hear back from you, we will assume your problem has been resolved and we will close this issue.

**Comment by yusing** [comment]
Tor Browser does work outside of containers. And restart the container does not fix the issue that unable to establish connection every time I start the tor-browser container

**Comment by jotrocken** [comment]
Same problem here. Connection don't get established. Tor Browser outside of Docker works. Restarting didn't help.

**Comment by cenosito** [comment]
Shame here, docker container torbrowser not stablish connection

**Comment by NeoSlugman** [comment]
Hi ! 
Same issue here : locally on my laptop even on a VPS.

I tried to close the tab, but same problem after automatically re opened.

**Comment by lawndoc** [comment]
Could someone share their compose file or run command so I can attempt to reproduce the issue on my end? My deployment is working so far in my testing

**Comment by eleaner** [comment]
I am having the same issue but it connects without any problem if I don't use any mapping for /app/Browser/TorBrowser/Data/Tor

do you guys define persistence volume too?

**Comment by eleaner** [comment]
I actually attempted to build the image for version 13.0 and the result is exactly the same

**Comment by eleaner** [comment]
it seems to be working ok with persistence of /app/Browser/TorBrowser/Data/Browser
maybe there is something new/different with tor?

**Comment by TobiasDax** [comment]
so, i am currently trying to set this up on CasaOS as a custom container with a Raspberry Pi as hardware.

CasaOS does not provide that much info, just a UI to click together all the instructions. I can provide a screenshot but i don't think thats of much help.

After that failed i SSHd into my raspberry pi and rn
`sudo docker run -d -p 5800:5800 domistyle/tor-browser`

with the same effect. the VNC connection starts and i can see the tor browser but if i click onto connect it loads for a bit and then just stops with the loading bar being only 20% full.

Feel free to ask follow up questions.

Here is the output from docker logs

```
[init        ] container is starting...
[cont-env    ] loading container environment variables...
[cont-env    ] APP_NAME: loading...
[cont-env    ] DISPLAY: executing...
[cont-env    ] DISPLAY: terminated successfully.
[cont-env    ] DISPLAY: loading...
[cont-env    ] DOCKER_IMAGE_PLATFORM: loading...
[cont-env    ] GTK2_RC_FILES: executing...
[cont-env    ] GTK2_RC_FILES: terminated successfully.
[cont-env    ] GTK2_RC_FILES: loading...
[cont-env    ] GTK_THEME: executing...
[cont-env    ] GTK_THEME: terminated successfully.
[cont-env    ] GTK_THEME: loading...
[cont-env    ] HOME: loading...
[cont-env    ] QT_STYLE_OVERRIDE: executing...
[cont-env    ] QT_STYLE_OVERRIDE: terminated successfully.
[cont-env    ] QT_STYLE_OVERRIDE: loading...
[cont-env    ] TAKE_CONFIG_OWNERSHIP: loading...
[cont-env    ] XDG_CACHE_HOME: loading...
[cont-env    ] XDG_CONFIG_HOME: loading...
[cont-env    ] XDG_DATA_HOME: loading...
[cont-env    ] XDG_RUNTIME_DIR: loading...
[cont-env    ] XDG_STATE_HOME: loading...
[cont-env    ] container environment variables initialized.
[cont-secrets] loading container secrets...
[cont-secrets] container secrets loaded.
[cont-init   ] executing container initialization scripts...
[cont-init   ] 10-certs.sh: executing...
[cont-init   ] 10-certs.sh: terminated successfully.
[cont-init   ] 10-check-app-niceness.sh: executing...
[cont-init   ] 10-check-app-niceness.sh: terminated successfully.
[cont-init   ] 10-cjk-font.sh: executing...
[cont-init   ] 10-cjk-font.sh: terminated successfully.
[cont-init   ] 10-clean-logmonitor-states.sh: executing...
[cont-init   ] 10-clean-logmonitor-states.sh: terminated successfully.
[cont-init   ] 10-clean-tmp-dir.sh: executing...
[cont-init   ] 10-clean-tmp-dir.sh: terminated successfully.
[cont-init   ] 10-fontconfig-cache-dir.sh: executing...
[cont-init   ] 10-fontconfig-cache-dir.sh: terminated successfully.
[cont-init   ] 10-init-users.sh: executing...
[cont-init   ] 10-init-users.sh: terminated successfully.
[cont-init   ] 10-nginx.sh: executing...
[cont-init   ] 10-nginx.sh: terminated successfully.
[cont-init   ] 10-openbox.sh: executing...
[cont-init   ] 10-openbox.sh: terminated successfully.
[cont-init   ] 10-set-tmp-dir-perms.sh: executing...
[cont-init   ] 10-set-tmp-dir-perms.sh: terminated successfully.
[cont-init   ] 10-vnc-password.sh: executing...
[cont-init   ] 10-vnc-password.sh: terminated successfully.
[cont-init   ] 10-web-data.sh: executing...
[cont-init   ] 10-web-data.sh: terminated successfully.
[cont-init   ] 10-x11-unix.sh: executing...
[cont-init   ] 10-x11-unix.sh: terminated successfully.
[cont-init   ] 10-xdg-runtime-dir.sh: executing...
[cont-init   ] 10-xdg-runtime-dir.sh: terminated successfully.
[cont-init   ] 15-install-pkgs.sh: executing...
[cont-init   ] 15-install-pkgs.sh: terminated successfully.
[cont-init   ] 85-take-config-ownership.sh: executing...
[cont-init   ] 85-take-config-ownership.sh: terminated successfully.
[cont-init   ] 89-info.sh: executing...
    ╭――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――╮
    │                                                                      │
    │ Application:           Tor Browser                                   │
    │ Application Version:   n/a                                           │
    │ Docker Image Version:  n/a                                           │
    │ Docker Image Platform: linux/arm64                                   │
    │                                                                      │
    ╰――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――╯
[cont-init   ] 89-info.sh: terminated successfully.
[cont-init   ] all container initialization scripts executed.
[init        ] giving control to process supervisor.
[supervisor  ] loading services...
[supervisor  ] loading service 'default'...
[supervisor  ] loading service 'logrotate'...
[supervisor  ] loading service 'logmonitor'...
[supervisor  ] service 'logmonitor' is disabled.
[supervisor  ] loading service 'app'...
[supervisor  ] loading service 'gui'...
[supervisor  ] loading service 'xvnc'...
[supervisor  ] loading service 'openbox'...
[supervisor  ] loading service 'certsmonitor'...
[supervisor  ] service 'certsmonitor' is disabled.
[supervisor  ] loading service 'nginx'...
[supervisor  ] all services loaded.
[supervisor  ] starting services...
[supervisor  ] starting service 'xvnc'...
[xvnc        ] Xvnc TigerVNC 1.13.0 - built Mar  4 2023 00:46:34
[xvnc        ] Copyright (C) 1999-2022 TigerVNC Team and many others (see README.rst)
[xvnc        ] See https://www.tigervnc.org for information on TigerVNC.
[xvnc        ] Underlying X server release 12013000
[xvnc        ] Wed Nov  1 17:35:40 2023
[xvnc        ]  vncext:      VNC extension running!
[xvnc        ]  vncext:      Listening for VNC connections on /tmp/vnc.sock (mode 0660)
[xvnc        ]  vncext:      Listening for VNC connections on all interface(s), port 5900
[xvnc        ]  vncext:      created VNC server for screen 0
[supervisor  ] starting service 'openbox'...
[supervisor  ] starting service 'nginx'...
[nginx       ] Listening for HTTP connections on port 5800.
[xvnc        ] Wed Nov  1 17:35:41 2023
[xvnc        ]  Connections: accepted: /tmp/vnc.sock
[xvnc        ]  SConnection: Client needs protocol version 3.8
[xvnc        ]  SConnection: Client requests security type None(1)
[supervisor  ] starting service 'app'...
[xvnc        ]  VNCSConnST:  Server default pixel format depth 24 (32bpp) little-endian rgb888
[app         ] Configuring Tor browser
[app         ] Starting Tor browser
[xvnc        ] Wed Nov  1 17:35:42 2023
[xvnc        ]  VNCSConnST:  Client pixel format depth 24 (32bpp) little-endian bgr888
[xvnc        ]  ComparingUpdateTracker: 0 pixels in / 0 pixels out
[xvnc        ]  ComparingUpdateTracker: (1:nan ratio)
[supervisor  ] all services started.
```

**Comment by Bolex80** [comment]
Having the same issue here.
Used docker compose file, and the output of my logs:

          [cont-env    ] GTK_THEME: executing...
          [cont-env    ] GTK_THEME: terminated successfully.
          [cont-env    ] GTK_THEME: loading...
          [cont-env    ] HOME: loading...
          [cont-env    ] QT_STYLE_OVERRIDE: executing...
          [cont-env    ] QT_STYLE_OVERRIDE: terminated successfully.
          [cont-env    ] QT_STYLE_OVERRIDE: loading...
          [cont-env    ] TAKE_CONFIG_OWNERSHIP: loading...
          [cont-env    ] XDG_CACHE_HOME: loading...
          [cont-env    ] XDG_CONFIG_HOME: loading...
          [cont-env    ] XDG_DATA_HOME: loading...
          [cont-env    ] XDG_RUNTIME_DIR: loading...
          [cont-env    ] XDG_STATE_HOME: loading...
          [cont-env    ] container environment variables initialized.
          [cont-secrets] loading container secrets...
          [cont-secrets] container secrets loaded.
          [cont-init   ] executing container initialization scripts...
          [cont-init   ] 10-certs.sh: executing...
          [cont-init   ] 10-certs.sh: terminated successfully.
          [cont-init   ] 10-check-app-niceness.sh: executing...
          [cont-init   ] 10-check-app-niceness.sh: terminated successfully.
          [cont-init   ] 10-cjk-font.sh: executing...
          [cont-init   ] 10-cjk-font.sh: terminated successfully.
          [cont-init   ] 10-clean-logmonitor-states.sh: executing...
          [cont-init   ] 10-clean-logmonitor-states.sh: terminated successfully.
          [cont-init   ] 10-clean-tmp-dir.sh: executing...
          [cont-init   ] 10-clean-tmp-dir.sh: terminated successfully.
          [cont-init   ] 10-fontconfig-cache-dir.sh: executing...
          [cont-init   ] 10-fontconfig-cache-dir.sh: terminated successfully.
          [cont-init   ] 10-init-users.sh: executing...
          [cont-init   ] 10-init-users.sh: terminated successfully.
          [cont-init   ] 10-nginx.sh: executing...
          [cont-init   ] 10-nginx.sh: terminated successfully.
          [cont-init   ] 10-openbox.sh: executing...
          [cont-init   ] 10-openbox.sh: terminated successfully.
          [cont-init   ] 10-set-tmp-dir-perms.sh: executing...
          [cont-init   ] 10-set-tmp-dir-perms.sh: terminated successfully.
          [cont-init   ] 10-vnc-password.sh: executing...
          [cont-init   ] 10-vnc-password.sh: terminated successfully.
          [cont-init   ] 10-web-data.sh: executing...
          [cont-init   ] 10-web-data.sh: terminated successfully.
          [cont-init   ] 10-x11-unix.sh: executing...
          [cont-init   ] 10-x11-unix.sh: terminated successfully.
          [cont-init   ] 10-xdg-runtime-dir.sh: executing...
          [cont-init   ] 10-xdg-runtime-dir.sh: terminated successfully.
          [cont-init   ] 15-install-pkgs.sh: executing...
          [cont-init   ] 15-install-pkgs.sh: terminated successfully.
          [cont-init   ] 85-take-config-ownership.sh: executing...
          [cont-init   ] 85-take-config-ownership.sh: terminated successfully.
          [cont-init   ] 89-info.sh: executing...
              ╭――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――╮
              │                                                                      │
              │ Application:           Tor Browser                                   │
              │ Application Version:   n/a                                           │
              │ Docker Image Version:  n/a                                           │
              │ Docker Image Platform: linux/amd64                                   │
              │                                                                      │
              ╰――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――╯
          [cont-init   ] 89-info.sh: terminated successfully.
          [cont-init   ] all container initialization scripts executed.
          [init        ] giving control to process supervisor.
          [supervisor  ] loading services...
          [supervisor  ] loading service 'default'...
          [supervisor  ] loading service 'logmonitor'...
          [supervisor  ] service 'logmonitor' is disabled.
          [supervisor  ] loading service 'logrotate'...
          [supervisor  ] loading service 'app'...
          [supervisor  ] loading service 'gui'...
          [supervisor  ] loading service 'certsmonitor'...
          [supervisor  ] service 'certsmonitor' is disabled.
          [supervisor  ] loading service 'xvnc'...
          [supervisor  ] loading service 'nginx'...
          [supervisor  ] loading service 'openbox'...
          [supervisor  ] all services loaded.
          [supervisor  ] starting services...
          [supervisor  ] starting service 'xvnc'...
          [xvnc        ] Xvnc TigerVNC 1.13.0 - built Mar  4 2023 00:46:41
          [xvnc        ] Copyright (C) 1999-2022 TigerVNC Team and many others (see README.rst)
          [xvnc        ] See https://www.tigervnc.org for information on TigerVNC.
          [xvnc        ] Underlying X server release 12013000
          [xvnc        ] Fri Nov  3 16:00:19 2023
          [xvnc        ]  vncext:      VNC extension running!
          [xvnc        ]  vncext:      Listening for VNC connections on /tmp/vnc.sock (mode 0660)
          [xvnc        ]  vncext:      Listening for VNC connections on all interface(s), port 5900
          [xvnc        ]  vncext:      created VNC server for screen 0
          [supervisor  ] starting service 'nginx'...
          [nginx       ] Listening for HTTP connections on port 5800.
          [supervisor  ] starting service 'openbox'...
          [supervisor  ] starting service 'app'...
          [app         ] Configuring Tor browser
          [app         ] Starting Tor browser
          [supervisor  ] all services started.
          [xvnc        ] Fri Nov  3 16:00:51 2023
          [xvnc        ]  Connections: accepted: /tmp/vnc.sock
          [xvnc        ]  SConnection: Client needs protocol version 3.8
          [xvnc        ]  SConnection: Client requests security type None(1)
          [xvnc        ]  VNCSConnST:  Server default pixel format depth 24 (32bpp) little-endian rgb888
          [xvnc        ]  VNCSConnST:  Client pixel format depth 24 (32bpp) little-endian bgr888
          [xvnc        ]  ComparingUpdateTracker: 0 pixels in / 0 pixels out
          [xvnc        ]  ComparingUpdateTracker: (1:-nan ratio)

I have clicked on Connect, but it doesnt do anything else:
![image](https://github.com/DomiStyle/docker-tor-browser/assets/8085823/eb284b7c-95e9-44da-ae56-02a864613899)

Now I can't even Cancell. It is non-responsive. 
I have tried closing the tabs, restarting, but nothing. No way to connect.

**Comment by lawndoc** [comment]
> so, i am currently trying to set this up on CasaOS as a custom container with a Raspberry Pi as hardware.

What are the specs of your RPi? Could you show the output of `free -h` on your RPi when it's frozen? I wonder if you're running into memory constraints.

I have an RPi in my closet I can test on, it'll take a little bit to get set up with Casa OS. I'll update with progress tomorrow.

**Comment by lawndoc** [comment]
@Bolex80 can you share your compose file please? Is this also running on a Raspberry Pi?

Also, please run `free -h` on the system when the browser is frozen to validate that it's not an OoM issue.

**Comment by eleaner** [comment]
I was running it on x86
as soon as the tor folder is mapped it hangs

**Comment by Bolex80** [comment]
> @Bolex80 can you share your compose file please? Is this also running on a Raspberry Pi?
> 
> Also, please run `free -h` on the system when the browser is frozen to validate that it's not an OoM issue.

It wasnt running on raspberry, I am using an ubuntu server x86.
I literaly used the docker-compose file from this link: https://github.com/DomiStyle/docker-tor-browser/blob/master/docker-compose.yml
I just changed the timezone and added the DARK_MODE variable. 

I have deleted it now, but will re-do it later to test the: free -h command.
I'll get back to you once I manage to do it.

Thanks anyway :)

**Comment by lawndoc** [comment]
Update -- I was able to reproduce the issue on my end. I'll do some troubleshooting and see if I can find the root cause. I'll provide another update no later than 2 days from now.

**Comment by lawndoc** [comment]
It definitely has something to do with the permissions of the mounted folder. By default, the container will create the volume with `root:root` as the owner, but then start the Tor Browser service using `app:app` with uid 1000. This causes the mounted folder to be unwritable by the app so it is empty. Then the browser can't function because of this.

There seems to be more going on than just the owner though, because changing the owner of mounted volume to `1000:1000` will only populate 2 of the 9 missing items in that folder.

I'll keep looking into the best way to mount a volume for this container, but for now I recommend running it without any mounted volumes.

**Comment by lawndoc** [comment]
@TobiasDax issue seems to be different because there is no volume mount in their command. That one may be arm-specific. I'll troubleshoot that separately.

**Comment by TobiasDax** [comment]
just to give you a bit more info, this is my `free -h` output while the container is running

```
~ $ free -h
               total        used        free      shared  buff/cache   available
Mem:           3.7Gi       1.9Gi       136Mi        74Mi       1.8Gi       1.8Gi
Swap:           99Mi       2.5Mi        97Mi
```

**Comment by lawndoc** [comment]
Thanks, definitely not a memory issue. I've come down with a cold or something the last couple of days so I haven't dug into this more since my last update. I'll find some time over the weekend to dig into this more and provide another update before Monday morning.

**Comment by lawndoc** [comment]
Checking back in, I'll be digging into this more tonight and will provide an update either way.

**Comment by lawndoc** [comment]
Didn't solve the issue last night. Gonna chip away at it when I have some free time (probably Sunday). If anyone else finds anything useful before then lmk

**Comment by lawndoc** [comment]
After reviewing the issue and taking a step back to think, I've decided to remove support for mounted volumes for 3 reasons:
- Persistent storage isn't great for privacy with Tor Browser usage, so discouraging persistent volumes protects our users
- The persistent storage previously recommended did not actually add any noticable speed increase to container startup
- Discouraging the use of mounted volumes increases the stability of container deployments

To summarize: __Moving forward, we recommend that you do not use persistent storage. If you do and it breaks something, we will not spend time supporting the issue.__

The `docker-compose.yml` file provided by this repository will be updated to remove the persistent storage, and the README will contain a message about not supporting volumes as well.

I apologize to anyone for not resolving the original issue, but I think this is the better way to go long-term. I welcome feedback if you think that there is a real need for volume support that doesn't majorly conflict with the 3 reasons I mentioned above.

**Comment by eleaner** [comment]
What about bookmarks?I doubt they impact privacy but are quite useful You need to map storage for that Sent from my iPhoneOn 21 Nov 2023, at 04:48, notifications at github.com ***@***.***> wrote:﻿
After reviewing the issue and taking a step back to think, I've decided to remove support for mounted volumes for 3 reasons:

Persistent storage isn't great for privacy with Tor Browser usage, so discouraging persistent volumes protects our users
The persistent storage previously recommended did not actually add any noticable speed increase to container startup
Discouraging the use of mounted volumes increases the stability of container deployments

To summarize: Moving forward, we recommend that you do not use persistent storage. If you do and it breaks something, we will not spend time supporting the issue.
The docker-compose.yml file provided by this repository will be updated to remove the persistent storage, and the README will contain a message about not supporting volumes as well.
I apologize to anyone for not resolving the original issue, but I think this is the better way to go long-term. I welcome feedback if you think that there is a real need for volume support that doesn't majorly conflict with the 3 reasons I mentioned above.

**Comment by lawndoc** [comment]
Bookmarks are a safe use case, you're right. I want to clarify though -- just because we don't support mounting volumes doesn't mean you can't do it anyways.

We are basically just saying that we won't normally spend dev time on issues related to mounted volumes.
